content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -5,6 +5,79 @@ prev: /recipes/authoring/kata-snippets/sample-tests/
+
+- If possible, **reference solution should not be used**. Sometimes there's no other way, but quite often reference solution is simply not necessary. Test cases can be generated with the answer known upfront, and when there's no reference solution to be run, it can save a lot of kinds of problems: time which would be used for running it is saved, you do not have to worry about it being accessible to users, or that it could be affected by mutation of the input. Sometimes, reference solutions happen to be bugged and tests fail even if user's solution is correct.
+- **Reference solution should not be revealed to the user.** When an assertion fails or the test suite crashes, some testing frameworks print fragments of source code which caused the failure to the console. It may happen that such printed failure messages or stack traces expose information about the solution which should not be revealed.
+- **Reference solution shouldn't be accessible to user solution.** It should not be possible to implement user solution as an alias or wrapper around reference solution used by tests. Reference solution should not be possible to be called by any other part of the kata other than the tests. In particular, it should not be a global function.  
+Chek the [reference page for your language](/languages/) to see how to prevent this problem in your tests.
+
+
+## Fixed Tests
+
+Fixed tests are tests with predetermined inputs, and do not change between test runs.
+
+- **Each requirement should have a corresponding fixed test.** Every requirement in the kata description should be explicitly tested with at least one, possibly more, dedicated and properly labeled assertion(s).
+- Tests should check not only valid inputs, but **invalid inputs and edge conditions should be tested too** (if problem allows them). `null` inputs, empty arrays, empty strings, etc. all should be tested with fixed tests, unless kata description explicitly says that such inputs won't be provided and don't have to be handled by a solution. Random tests are not a replacement for carefully designed assertions for edge cases.
+- If your test suite uses random tests (and it usually should, see below), then make sure that every test scenario generated there has its equivalent fixed test. Fixed tests are usually easier to debug, so ideally a fixed test should fail before corresponding scenario is tested with random tests.
+
+
+## Random Tests
+
+Random tests are not common testing practice in "real life" coding, and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hardcoding, and other workarounds. Goal of random tests is to make input of test cases unpredictable, so only solutions which actually solve the problem can pass.
```

Comment:
```suggestion
Random tests are uncommon in "real life" coding and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hard-coding, and other workarounds. The goal of random tests is to make the inputs unpredictable so that only proper solutions may pass.
```